common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py

```diff
@@ -1,50 +1,153 @@
 """
 Segregation of pymongo functions from the data modeling mechanisms for split modulestore.
 """
-import re
-from mongodb_proxy import autoretry_read, MongoProxy
+import datetime
+import math
 import pymongo
+import pytz
+import re
+from contextlib import contextmanager
+from time import time
 
 # Import this just to export it
 from pymongo.errors import DuplicateKeyError  # pylint: disable=unused-import
 
 from contracts import check, new_contract
+from mongodb_proxy import autoretry_read, MongoProxy
 from xmodule.exceptions import HeartbeatFailure
 from xmodule.modulestore import BlockData
 from xmodule.modulestore.split_mongo import BlockKey
-import datetime
-import pytz
+import dogstats_wrapper as dog_stats_api
 
 
 new_contract('BlockData', BlockData)
 
 
-def structure_from_mongo(structure):
+def round_power_2(value):
+    """
+    Return value rounded up to the nearest power of 2.
+    """
+    if value == 0:
+        return 0
+
+    return math.pow(2, math.ceil(math.log(value, 2)))
+
+
+class Tagger(object):
+    def __init__(self):
+        self.added_tags = []
+        self.measures = []
+
+    def measure(self, name, size):
+        self.measures.append((name, size))
+
+    def tag(self, **kwargs):
+        self.added_tags.extend(kwargs.items())
+
+    @property
+    def tags(self):
+        return [
+            '{}:{}'.format(name, round_power_2(size))
+            for name, size in self.measures
+        ] + [
+            '{}:{}'.format(name, value)
+            for name, value in self.added_tags
+        ]
+
+
+class QueryTimer(object):
+    def __init__(self, metric_base, sample_rate=1):
+        """
+        Arguments:
+            metric_base: The prefix to be used for all queries captured
+            with this :class:`QueryTimer`.
+        """
+        self._metric_base = metric_base
+        self._sample_rate = sample_rate
+
+    @contextmanager
+    def timer(metric_name, course_context):
+        """
```

Comment:
Since you mostly use the method name below, would it be worth finding the calling method name on the stack here for `metric_name`? Or would that have performance concerns?